### PR TITLE
documentation: configuration switch for an odoc documentation mode

### PR DESCRIPTION
**Pull request by Octachron**
This PR builds upon #9996 and adds an odoc documentation mode to the documentation build process.
This mode is hidden behind a configuration switch `--enable-odoc-documentation`.

When this switch is active, the documentation process uses odoc for generating the documentation for both the standalone documentation and the manual.

Note that this requires a master version of odoc compiled with 4.12.0 (and https://github.com/ocaml/odoc/pull/526 might be required or not depending on the phases of Saturn moons)

The standalone documentation can also be generated with `make -C doc odoc-all`.

### CODE REVIEW

File: configure.ac
Comment (gasche) [inline configure.ac]: the help string looks like a copy/paste error

**Review by gasche — COMMENTED** [review]
I only read this very quickly.

For the configure interface, I am wondering if it wouldn't be nicer for the user to directly choose the "documentation generator" as either `ocamldoc` or  `odoc`. The interface where `--enable-foo` changes a default, but in fact both foo and not-foo can stil be built on explicit demand, is slightly more surprising. (But then I guess other parts of the compiler use this, so maybe this is fine.)

Out of curiosity, what happens if I try to build the target `make odoc-all` after a non-odoc run of the documentation? (Or the other way around?) Does everything explode, or does it work fine? If the former, then maybe we want to be a bit more strict in requiring the configuration setting to let people build.

**Comment by Octachron** [comment]
The odoc-% and ocamldoc-% targets are orthogonal from the default one, there are mostly here as a quick test option.

**Review by gasche — APPROVED** [review]
This looks good, it only needs rebasing once #9996 is merged.

**Comment by shindere** [comment]
Will do a review, thanks.

Re: the configure option, I'd prefer `--with-odoc`. It'd be closer
to the intended semantics of ``--with`` vs. ``--enable`` as documented
in the autoconf manual.

**Comment by shindere** [comment]
ôOctachron: do you still have the wish to get this merged?

I started reviewing but found that I havd comments a bit similar to those
I made for #10191.

Would you mind refreshing your PR by taking into account the comments
I had done there?

If you agree, I'd be very happy to review the present PR. Thanks!

**Comment by gasche** [comment]
@shindere are you sure that you want to review this PR? I believe that it is okay to merge as-is (I didn't look in the patch in full details, but I think that for the manual we can be a bit less careful, and in particular the fact that the build still works provides good feedback on the PR), so if you are busy with other things we could just go ahead. If you believe it's useful for you and the code to do a review, then of course there is no hurry.

**Comment by shindere** [comment]
I gave some off-line feedback to @Octachron on this.

I'll leave it up to him to decide whether he wants to integrate them in
this PR or not.

**Comment by Octachron** [comment]
I am finalizing an updated version of this PR integrating @shindere 's feedback to have some more structure in those Makefiles.

**Comment by shindere** [comment]
Thanks, @Octachron!

Will you please make sure that the description of the directory structure for the output files given in `manual/README` matches your changes?

I will be on vacation next week (from March 8 to March 12) but would be
happy to review this before it gets merged, once I am back.

Please make sure to write a comment here when you are happy with your
changes and think time has come for a review!

File: .gitignore
Comment (dra27) [inline .gitignore]: Why are these moving to a separate `.gitignore`? For better or worse, the policy is a top-level one.

File: api_docgen/Makefile
Comment (dra27) [inline api_docgen/Makefile:22]:
```suggestion
include $(DOCUMENTATION_TOOL)/Makefile
```
?

File: configure.ac
Comment (dra27) [inline configure.ac:140]:
```suggestion
AC_SUBST([documentation_tool])
AC_SUBST([documentation_tool_cmd])
```
(see below)

File: configure.ac
Comment (dra27) [inline configure.ac]:
```suggestion
```
(see above)

File: configure.ac
Comment (dra27) [inline configure.ac:275]: All `--with` flags have an argument - in this case, it might be useful to support it properly to allow `--with-doc ~/odoc-dev/_build/...`

```suggestion
documentation_tool_cmd=''
AC_ARG_WITH([odoc],
  [AS_HELP_STRING([--with-odoc],
  [AS_CASE([$withval],
    [yes],[documentation_tool='odoc'],
    [no],[documentation_tool='ocamldoc'],
    [documentation_tool_cmd="$withval"
    documentation_tool='odoc'])],
  [documentation_tool='ocamldoc'])

AS_IF([test "x$documentation_tool_cmd" = 'x'],[documentation_tool_cmd="$documentation_tool"])
```
(the `[no]` case is what autoconf does if you explicitly say `--without-odoc` vs the final parameter to `AC_ARG_WITH` which is what's used if you don't specify either `--with-odoc` or `--without-odoc`)

File: Makefile.config.in
Comment (dra27) [inline Makefile.config.in]:
```suggestion
DOCUMENTATION_TOOL=@documentation_tool@
DOCUMENTATION_TOOL_CMD=@documentation_tool_cmd@
```
Regardless of this change, please could this go in `Makefile.build_config` - the (slow) aim is that `Makefile.config` (which is installed) should only contain things which may be of interest post-installation.

File: api_docgen/odoc/Makefile
Comment (dra27) [inline api_docgen/odoc/Makefile]:
```suggestion
odoc=$(DOCUMENTATION_TOOL_CMD)
```

File: api_docgen/odoc/Makefile
Comment (dra27) [inline api_docgen/odoc/Makefile:184]:
```suggestion
INSTALL_MAN3DIR=$(INSTALL_MANDIR)/man3
```
? The previous name looked like a typo?

File: api_docgen/ocamldoc/Makefile
Comment (dra27) [inline api_docgen/ocamldoc/Makefile:120]:
```suggestion
INSTALL_MAN3DIR=$(INSTALL_MANDIR)/man3
```
(with equivalent change below and see comment in the odoc `Makefile`).

**Review by dra27 — COMMENTED** [review]
I haven't done a particularly detailed review - a few suggestions/requests:
- The new variable(s) in `Makefile.config` aren't needed after installation
- I think when working on odoc it may well be useful to be able to customise the command being called rather than having to muck around with `PATH`, so I propose a bit `m4sh` for dealing with `--with-odoc=` pedantically (the existing code is fine except for the pedantically strange `--with-odoc=odoc` selecting `ocamldoc`!)
- Not sure why `.gitignore` was splitting out from the toplevel one

File: api_docgen/Makefile
Comment (Octachron) [reply api_docgen/Makefile:22]: This feels a bit too accidental.

**Comment by Naereen** [comment]
Hello there!
I am not an expert of odoc, but is it able to handle multiple localizations of the same documentation?
Sphinx for Python can do that, and since a few years the Python officiel documentation is available in a few other languages (e.g. <https://docs.python.org/fr/3/>). **I'm extra motivated to help to translate OCaml documentation to French!**

**Comment by Octachron** [comment]
As far as I know internationalization is not (yet?) on the odoc roadmap. For such an odoc feature wish, it is better to have a look at https://github.com/ocaml/odoc/issues and open an issue if needed.

**Comment by gasche** [comment]
I wondered where we are with this PR. The current status is that @dra27 has done (about a month ago) a more detailed review than mine, pointing out a moderate number of issues in the build system and configuration layer; @Octachron has not yet updated the PR accordindgly.

File: api_docgen/odoc/Makefile
Comment (Octachron) [reply api_docgen/odoc/Makefile:184]: Maybe, but this would be a mostly unrelated change?

File: api_docgen/odoc/Makefile
Comment (Octachron) [reply api_docgen/odoc/Makefile]: (Done with some more handling to the case where we are in the ocamldoc mode)

File: Makefile.config.in
Comment (Octachron) [reply Makefile.config.in]: I have moved those variables to `Makefile.build_config`, thanks for the suggestion.

File: configure.ac
Comment (Octachron) [reply configure.ac:275]: Updated.

**Comment by Octachron** [comment]
@shindere: this should be ready to be re-reviewed.

**Comment by Octachron** [comment]
@gasche , the PR has been updated now (It was stalled for a time because odoc needed to be updated for 4.13).

File: api_docgen/odoc/Makefile
Comment (dra27) [reply api_docgen/odoc/Makefile:184]: Sort of - though this _addition_ looked like a typo when I reviewed it (even if the addition is the result of moving code from somewhere else)

**Comment by shindere** [comment]
Thanks a lot for your hard work on this PR, @Octachron.

There are variable definitions where a space could be added before the =
sign, and stuff like that, but it's probably not very constructive to
nitpick on these things. Perhaps the best way to go is to get thismerged
and improve from that, which is already a good basis.

**Review by shindere — APPROVED** [review]

File: Changes
Comment (gasche) [inline Changes]: ? has to be instantiated here

**Comment by shindere** [comment]
You'll perhaps want to cleanup the commit history and complete che
Changes entry.

**Comment by Octachron** [comment]
I have updated the list of reviewers. Concerning the intermediary commits, I am planning to squash them all together.

**Comment by shindere** [comment]
Fine for the commits, thanks!

**Pull request merged**
